Extend TLS Support (#8505)
netdata/netdata: web/server/README.md

openssl speed rsa2048 rsa4096
```

### Select TLS version

> Since version 1.21 Netdata accepts `tls key` and `tls certificate` as replacement of the options `ssl key` and `ssl certificate`, 

Reviewer: ```suggestion
Beginning with version 1.21, Netdata accepts `tls key` and `tls certificate` as replacement of the options `ssl key` and `ssl certificate`, 
```

>  but you do not need to change your `netdata.conf` case you are using old `ssl` variables, because old options are kept due compatibility.

Reviewer: ```suggestion
Beginning with version 1.21, Netdata accepts `tls key` and `tls certificate` as replacement of the settings `ssl key` and `ssl certificate`. 
You do not need to change your `netdata.conf` if you are using the previous `ssl` settings, because these are kept for backwards compatibility.
```

Author: I used all suggestions that you gave me, thank you!

> Netdata also brings two new options related to TLS, now you can specify the TLS version and the ciphers that you want to use:

Reviewer: ```suggestion
You can also specify the TLS version and the ciphers that you want to use:
```

Author: I will mark this and the next as resolved, for the motive I already given.

```conf
[web]
    tls version = 1.3
    tls ciphers = TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256
```

> case you do not specify these options, Netdata will use the highest options available on your system.

Reviewer: ```suggestion
If you do not specify these options, Netdata will use the highest options available on your system.
```

An important note about the `tls version`, Netdata accepts all the TLS version as arguments (`1`, `1.1`, `1.2` and `1.3`), but

>  only the last option is considered safe.

Reviewer: ```suggestion
> While Netdata accepts all TLS versions as arguments (`1`, `1.1`, `1.2` and `1.3`), we recommend you use `1.3` for the most secure encryption.
```

### Select TLS version

Beginning with version 1.21, Netdata accepts `tls key` and `tls certificate` as replacement of the settings `ssl key` and `ssl certificate`. 

You do not need to change your `netdata.conf` if you are using the previous `ssl` settings, because these are kept for backward compatibility.

You can also specify the TLS version and the ciphers that you want to use:

```conf
[web]
    tls version = 1.3
    tls ciphers = TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256
```

> If you do not specify these options, Netdata will use the highest options available on your system.

Reviewer 2: For at least the ciphers, this isn't true. Netdata will use the default cipher list based on the selected protocol, which may not be the most secure possible.

Author: Case we use the highest protocol version, 1.3 right now, we will discard a lot of ciphers that are not considered safe.

Reviewer 2: That doesn't mean we automatically use the most secure ciphers though, which is what the statement above implies.. Take a look at the standard list for 1.2 for example, almost half of them are considered to at best be weak and at worst completely broken.

I'd suggest the following alternative:
```suggestion
If you do not specify these options, Netdata will use the highest available protocol version on your system and the default cipher list for that protocol provided by your TLS implementation.
```

Author: I accepted the suggestion, I am finishing the tests here, and I will commit everything in few minutes. Thank you!

Author: Done!

> While Netdata accepts all the TLS version as arguments (`1`, `1.1`, `1.2` and `1.3`), we recommend you use `1.3` for the most secure encryption.

Reviewer 2: Two things WRT this that may need to be considered in the code:

* We should probably accept 1.0 and alias it to 1 for the protocol version, as that's what most other software does.
* It would be a very good idea to support a list of protocol versions instead of enforcing a specific one. Only allowing a specific protocol is potentially problematic from the perspective of client-side support (that is, it may not be possible to support all the clients a user wants to support if they are only able to select a single protocol).

Author: For OpenSSL we have only SSLv3 and TLSx.y and considering that the option is named `tls version` I think this is not a problem, case OpenSSL uses another protocol, we will have to deal with this. 
About the alias, I think it is a good point, I will bring it in few minutes. Thank you!

Reviewer 2: There are enough differences though between TLS 1.2 and 1.3 that clients that only support up to 1.2 can't connect to a server that only supports 1.3. That's kind of the reason that there's a version distinction there.

And, unfortunately, there are still a lot of systems out there with clients that only support 1.2, or even just 1.1, because anything newer required a new version of OpenSSL with some rather problematic ABI compatibility issues.

Author: I understand this, among the tests that I did, I forced the `slave` to use `1.2` and the server with `1.3` and as expected it could not work. 

I did not add any additional information for the issue about the relationship between tls versions, because openssl always try to get the highest value between client and server, on the other hand, I was forcing to use an unique version of the protocol. I will let this flexible between the minimum and maximum available for our users do not have problems.
